### PR TITLE
Implement C/C++ PairwiseANN w/ Python API

### PR DESCRIPTION
**Pull request by OctoberChang**
*Issue #, if available:*

*Description of changes:*
Implement C/C++ PairwiseANN function with Python API.

Given a (input, label) pair, PairwiseANN finds top-K nearest queries in the indexed input-to-label graph.
The searcher return four arrays of size `K`:
- `Imat`: top-K input indices
- `Mmat`: the masking indicators. 1 indicates the kNN index is presented; otherwise 0. 
- `Dmat`: the kNN distances between the test input and top-K indexed input.
- `Vmat`: the corresponding values stored in the input-to-label graph.

*Indexing usage*
```
train_params = PairwiseANN.TrainParams(metric_type="ip")
model = PairwiseANN.train(X_trn, Y_csr, train_params=train_params)
```
where
- `X_trn` is the dense or sparse input feature matrix
- `Y_csr` is the sparse input-to-label graph

*Save/load Usage*
```
model.save(model_folder)
del model
model = PairwiseANN.load(model_folder)
```

*Searchers Usage*
```
searchers = model.searchers_create(max_batch_size=100, max_only_topk=10, num_searcher=1)
```
where
- `max_batch_size` defines the maximum number of rows in the returned array
- `max_only_topk` defines the maximum number of cols in the returned array

*Prediction Usage*
With same input (suitable for real-time inference):
```
It, Mt, Dt, Vt = model.predict(query_vec, label_keys, searchers, pred_params=pred_params, is_same_input=True)
```
- `query_vec` is a single feature vector of shape `(1, feat_dim)`
- `label_keys` is a np.array of shape `(batch_size, )`

With different inputs (suitable for batch prediction):
```
It, Mt, Dt, Vt = model.predict(Query_mat, label_keys, searchers, pred_params=pred_params, is_same_input=False)
```
- `Query_mat` is a batch feature matrix of shape `(batch_size, feat_dim)`
- `label_keys` is a np.array of shape `(batch_size, )`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/ann/pairwise/model.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            data_type = "csr"
+        else:
+            raise ValueError("type(X)={} is NOT supported!".format(type(X)))
+        return pX, data_type
+
+    @classmethod
+    def train(cls, X, Y, train_params=None, pred_params=None):
+        """train and return the PairwiseANN indexer
+        Args:
+            X (numpy.array or smat.csr_matrix): database matrix to be indexed. (num_item x feat_dim).
```
Comment (jiong-zhang) [inline pecos/ann/pairwise/model.py]: Missing `Y` in docstring.

File: pecos/ann/pairwise/model.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            self.Dmat[:reset_nnz].fill(0.0)
+            self.Vmat[:reset_nnz].fill(0.0)
+
+    def __init__(
+        self, model_ptr, num_input_keys, num_label_keys, feat_dim, fn_dict, pred_params=None
+    ):
+        """constructor of PairwiseANN class
+        Args:
+            model_ptr (c_void_p): pointer to C instance pecos::ann:PairwiseANN.
+            num_item (int): number of item being indexed
```
Comment (jiong-zhang) [inline pecos/ann/pairwise/model.py]: Please fix docstring.

File: pecos/core/ann/pairwise.hpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    template<class FeatVec_T, class MAT_T>
+    struct PairwiseANN {
+        typedef FeatVec_T feat_vec_t;
+        typedef MAT_T mat_t;
+        typedef pecos::ann::KeyValPair<index_type, value_type> pair_t;
+        typedef pecos::ann::heap_t<pair_t, std::less<pair_t>> max_heap_t;
+
+		struct Searcher {
+            typedef PairwiseANN<feat_vec_t, mat_t> pairwise_ann_t;
+            typedef pecos::ann::heap_t<pair_t, std::less<pair_t>> max_heap_t;
```
Comment (jiong-zhang) [inline pecos/core/ann/pairwise.hpp]: Don't need to define this again.

File: pecos/core/ann/pairwise.hpp [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+ 
+	// PairwiseANN Interface
+    template<class FeatVec_T, class MAT_T>
+    struct PairwiseANN {
+        typedef FeatVec_T feat_vec_t;
+        typedef MAT_T mat_t;
+        typedef pecos::ann::KeyValPair<index_type, value_type> pair_t;
+        typedef pecos::ann::heap_t<pair_t, std::less<pair_t>> max_heap_t;
+
+		struct Searcher {
```
Comment (jiong-zhang) [inline pecos/core/ann/pairwise.hpp]: Indentation

File: pecos/core/ann/pairwise.hpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+                // load matrices
+                load_mat<pecos::csc_t>(Y_csc, mmap_store);
+                load_mat<mat_t>(X_trn, mmap_store);
+                // DO NOT call mmap_store.close() as the actual memory is held by this->mmap_store object.
+            } else {
+                throw std::runtime_error("Unable to load memory-mapped file with version = " + version);
+            }
+        }
+
+        void train(const mat_t &X_trn, const pecos::csr_t &Y_trn) {
```
Comment (jiong-zhang) [inline pecos/core/ann/pairwise.hpp]: should we just assume `Y_trn` to be CSC here? We can support CSR/CSC at python level.

File: pecos/core/libpecos.cpp [hunk 6]
```diff
@@ -476,6 +477,102 @@ extern "C" {
...
+        void* searchers_ptr, \
+        uint32_t batch_size, \
+        uint32_t topk, \
+        const PY_MAT* pQ, \
+        uint32_t* label_keys, \
+        uint32_t* ret_Imat, \
+        uint32_t* ret_Mmat, \
+        float* ret_Dmat, \
+        float* ret_Vmat, \
+        const bool is_sample_input) { \
```
Comment (jiong-zhang) [inline pecos/core/libpecos.cpp]: Should this be `is_same_input`?

File: pecos/ann/pairwise/model.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+    def get_pred_params(self):
+        """Return a deep copy of prediction parameters
+
+        Returns:
+            copied_pred_params (dict): Prediction parameters.
+        """
+        return copy.deepcopy(self.pred_params)
+
+    def predict(self, input_feat, label_keys, searchers, pred_params=None, is_same_input=False):
```
Comment (jiong-zhang) [inline pecos/ann/pairwise/model.py:240]: I wonder if there's a more extendable design for the predict API. Now we are basically supporting two cases (1) `is_samp_input=True`, a csr_vector and (2) `is_same_input=False`, a coo matrix.

File: pecos/ann/pairwise/model.py (hunk 2 above)
Comment (OctoberChang) [reply pecos/ann/pairwise/model.py]: Fixed. See Revision 2.

File: pecos/core/ann/pairwise.hpp (hunk 3 above)
Comment (OctoberChang) [reply pecos/core/ann/pairwise.hpp]: Fixed. See Revision 2.

File: pecos/ann/pairwise/model.py (hunk 1 above)
Comment (OctoberChang) [reply pecos/ann/pairwise/model.py]: Fixed. See Revision 2.

File: pecos/core/ann/pairwise.hpp (hunk 4 above)
Comment (OctoberChang) [reply pecos/core/ann/pairwise.hpp]: Fixed. See Revision 2.

File: pecos/core/ann/pairwise.hpp (hunk 5 above)
Comment (OctoberChang) [reply pecos/core/ann/pairwise.hpp]: Fixed. See Revision 2.

File: pecos/core/libpecos.cpp (hunk 6 above)
Comment (OctoberChang) [reply pecos/core/libpecos.cpp]: Fixed. See Revision 2.

**Review by jiong-zhang — APPROVED** [review]

**Pull request merged**